DOCSP-41963: Databases and collections (#136)
mongodb/docs-php-library: source/databases-collections.txt

  parameters
- ``MongoDB\Database::selectCollection()``: Pass the collection name as a parameter

The following example accesses a collection named ``test_collection`` by using the 
``MongoDB\Database::selectCollection()`` method:

.. literalinclude:: /includes/databases-collections/databases-collections.php
    :language: php
    :dedent:
    :start-after: start-access-collection
    :end-before: end-access-collection

.. tip::

   If the provided collection name does not already exist in the database,
   MongoDB implicitly creates the collection when you first insert data
   into it.

.. _php-db-coll-create-collection:

Create a Collection
-------------------

Pass a collection name to the ``MongoDB\Database::createCollection()`` method to
explicitly create a collection in a MongoDB database.

The following example creates a collection named ``example_collection``:

.. literalinclude:: /includes/databases-collections/databases-collections.php
    :language: php
    :dedent:
    :start-after: start-create-collection
    :end-before: end-create-collection

You can specify collection options, such as maximum size and document

> validation rules, by passing them as an array parameter to ``createCollection()``.

Reviewer: S:

```suggestion
validation rules, by passing them as an array to the ``createCollection()`` method.
```

For a full list of optional parameters, see the `API documentation

> <{+api+}/method/MongoDBDatabase-createCollection/#parameters>`__.

Reviewer 2: I assume `:phpmethod: is not used here because you want to deep link to the parameters heading?

Author: yes!

Get a List of Collections
-------------------------

You can query for a list of collections in a database by calling the
``MongoDB\Database::listCollections()`` method. The method returns a
cursor containing all collections in the database and their associated metadata.

The following example calls the ``listCollections()`` method and iterates over

> the returned cursor to print the collections from the :ref:`php-db-coll-access-collection`

Reviewer 2: ```suggestion
the returned iterator to print the collections from the :ref:`php-db-coll-access-collection`
```

Although the [`listCollections` command](https://www.mongodb.com/docs/manual/reference/command/listCollections/) returns a cursor, PHPLIB abstracts that with another iterator. This allows us to do things like [cache the results](https://github.com/mongodb/mongo-php-library/blob/v1.x/src/Model/CachingIterator.php) and return typed model classes (e.g. CollectionInfo).

Caching is helpful if users want to iterate the result multiple times, as the original cursor can only be iterated once as it streams results from the server.

Also note that the special iterator classes returned by the database, collection, and index enumeration methods are being deprecated and will be removed in 2.0 (see: [PHPLIB-1522](https://jira.mongodb.org/browse/PHPLIB-1522)). The individual result classes (e.g. CollectionInfo) will remain, though. In order to future-proof the docs, I propose we just talk about these methods returning an iterator.

and :ref:`php-db-coll-create-collection` examples:

.. io-code-block::
   :copyable:

   .. input:: /includes/databases-collections/databases-collections.php
      :language: php
      :start-after: start-find-collections
      :end-before: end-find-collections
      :dedent:

   .. output::
      :language: console
      :visible: false

      MongoDB\Model\CollectionInfo Object
      (
         [name] => example_collection
         [type] => collection
         ...
      )
      MongoDB\Model\CollectionInfo Object
      (
         [name] => test_collection
         [type] => collection
         ...
      )

> For more information about iterating over a cursor, see the :ref:`php-cursors`
> guide. 

Reviewer 2: ```suggestion
```

This can be removed since none of these enumeration methods return a cursor.

Delete a Collection
-------------------

You can delete a collection from the database by using the ``MongoDB\Database::dropCollection()``
method.

The following example deletes the ``test_collection`` collection:

.. literalinclude:: /includes/databases-collections/databases-collections.php
    :language: php
    :dedent:
    :start-after: start-drop-collection
    :end-before: end-drop-collection

.. warning:: Dropping a Collection Deletes All Data in the Collection

   Dropping a collection from your database permanently deletes all
   documents and all indexes within that collection.

>    Drop a collection only if the data in it is no longer needed.

Reviewer: S:

```suggestion
   Drop a collection only if you no longer need the data in it.
```

.. _php-config-read-write:

Configure Read and Write Operations
-----------------------------------

You can control how the library routes read operations by setting a **read preference**.
You can also control options for how the library waits for acknowledgment of
read and write operations on a replica set by setting a **read concern** and a
**write concern**.

By default, databases inherit read and write settings from the ``MongoDB\Client``

> instance, and collections inherit them from the database. However, you can change

Reviewer: ```suggestion
instance, and collections inherit them from the database. You can change
```

these settings on your database by passing an options array to the
``MongoDB\Client::selectDatabase()``. You can change these settings on your collection

> by passing an options array to the ``MongoDB\Client::selectCollection()`` method.

Reviewer 2: Regarding collection inheritance, note that if users call `Client::selectCollection()` they would effectively inherit the Client's options since an interstitial Database object is never created.

`Database::selectCollection()` isn't even mentioned in this paragraph, although it was in a previous section. I'll note that the code example only demonstrates `Database::selectCollection()`, so perhaps you want to focus on that here instead of the Client method.

I'll defer to you if any of this requires some attention.

To change the read preference, read concern, and write concern of your database
or collection, set the following options in the array parameter:

- ``readPreference``: Sets the read preference. For a list of available read
  preferences, see `MongoDB\\Driver\\ReadPreference <{+php-manual+}/class.mongodb-driver-readpreference>`__
  in the extension API documentation.
- ``readConcern``: Sets the read concern. For a list of available read

>   concerns, see `MongoDB\\Driver\\ReadConcern <{+php-manual+}/class.mongodb-driver-readconcern>`__

Reviewer 2: Existing docs use the `:php:` directive. For example:

```
:php:`MongoDB\Driver\ReadConcern <class.mongodb-driver-readconcern>`
```

Are there plans to eliminate that? If not, should you use it for PHP.net links?

Author: Good to know, I wasn't aware of that directive - updated

  in the extension API documentation.
- ``writeConcern``: Sets the write concern. For a list of available write

>   concerns, see `MongoDB\\Driver\\ReadConcern <{+php-manual+}/class.mongodb-driver-writeconcern>`__

Reviewer: ```suggestion
  concerns, see `MongoDB\\Driver\\WriteConcern <{+php-manual+}/class.mongodb-driver-writeconcern>`__
```

  in the extension API documentation.

.. tip::

>    To learn more about the read and write settings, see the following guides in the

Reviewer: ```suggestion
   To learn more about read preferences and read and write concerns, see the following guides in the
```

   MongoDB Server manual:

   - :manual:`Read Preference </core/read-preference/>`
   - :manual:`Read Concern </reference/read-concern/>`
   - :manual:`Write Concern </reference/write-concern/>`

Database Configuration Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to set the read preference, read concern, and

> write preference of a database called ``test_database`` by passing an options

Reviewer: ```suggestion
write concern of a database called ``test_database`` by passing an options
```

array to ``selectDatabase()``:

.. literalinclude:: /includes/databases-collections/databases-collections.php
    :language: php
    :dedent:
    :start-after: start-database-settings
    :end-before: end-database-settings

Collection Configuration Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to set the read preference, read concern, and

> write preference of a collection called ``test_collection`` by passing an options

Reviewer: ```suggestion
write concern of a collection called ``test_collection`` by passing an options
```

array to ``selectCollection()``:

.. literalinclude:: /includes/databases-collections/databases-collections.php
    :language: php
    :dedent:
    :start-after: start-collection-settings
    :end-before: end-collection-settings

> Tag Sets

Reviewer 2: Should "Tag Sets" and "Local Threshold" be grouped under a "Read Preference" heading?

Reviewer 2: Noted that "Advanced Read Configurations" was added.

~~~~~~~~

In {+mdb-server+}, you can apply key-value :manual:`tags
</core/read-preference-tags/>` to replica-set
members according to any criteria you choose. You can then use
those tags to target one or more members for a read operation.

By default, the {+php-library+} ignores tags when choosing a member
to read from. To instruct the {+php-library+} to prefer certain tags,
pass them as a parameter to your ``MongoDB\Driver\ReadPreference`` class
constructor. Then, set the ``MongoDB\Driver\ReadPreference`` object as
the value of the ``readPreference`` database option.

The following code example sets the ``readPreference`` option to a tag set
that instructs ``test_database`` to prefer reads from the New York data
center (``'dc' => 'ny'``) and to fall back to the San Francisco data

> center (``'dc' => 'sf'``):

Reviewer 2: It's been a while since I revisited tag set docs, but I recall examples typically included an empty tag set as a fallback.

As-is, server selection will fail if there are no servers in either `dc:ny` or `dc:sf`. Just want to confirm that is intentional.

Author: Added an empty tag set

.. literalinclude:: /includes/databases-collections/databases-collections.php
    :language: php
    :dedent:
    :start-after: start-tag-set
    :end-before: end-tag-set

Local Threshold
~~~~~~~~~~~~~~~

If multiple replica-set members match the read preference and tag sets you specify,
the {+php-library+} reads from the nearest replica-set members, chosen according to
their ping time.

> By default, the library uses only those members whose ping times are within 15 milliseconds
> of the nearest member for queries. To distribute reads between members with
> higher latencies, pass an options array to the ``MongoDB\Client`` constructor that
> sets the ``localThresholdMS`` option.

Reviewer: S: this might be slightly clearer

```suggestion
The library uses only those members whose ping times are within the local
threshold of the nearest member for queries. By default,
the local threshold is 15 milliseconds of latency. To distribute reads among members with
higher latencies, pass an options array to the ``MongoDB\Client`` constructor that
sets the ``localThresholdMS`` option.
```

The following example specifies a local threshold of 35 milliseconds:

.. literalinclude:: /includes/databases-collections/databases-collections.php
    :language: php
    :dedent:
    :start-after: start-local-threshold
    :end-before: end-local-threshold

> In the preceding example, the {+php-library+} distributes reads between matching members

Reviewer: https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#a:~:text=00%20p.m.-,among,-between%20When%20referring

```suggestion
In the preceding example, the {+php-library+} distributes reads among matching members
```
